File: lambdas/algorithms/searching/java-searching/src/main/java/BinarySearch.java
public class BinarySearch extends BaseSearch {

    @Override
    protected int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            iterations++;
            int mid = left + (right - left) / 2;

            comparisons++; // nums[mid] == target
            if (nums[mid] == target) return mid;

            comparisons++; // nums[mid] < target
            if (nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }

        return -1;
    }

}

File: lambdas/algorithms/sorting/python-sorting/src/algorithms/binary_search.py
from src.base_search import BaseSearch
from typing import List

class BinarySearch(BaseSearch):
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            self.iterations += 1
            mid = left + (right - left) // 2

            self.comparisons += 1  # nums[mid] == target
            if nums[mid] == target:
                return mid

            self.comparisons += 1  # nums[mid] < target
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        return -1